numpy/core/src/npysort/quicksort.c.src

```diff
@@ -107,6 +111,7 @@ quicksort_@suff@(void *start, npy_intp num, void *NOT_USED)
                 *sptr++ = pi - 1;
                 pl = pi + 1;
             }
+            *psdepth++ = cdepth + 1;
```

Comment:
I'm not so sure about the performance impact of the recursion, I would not expect it to be large.
The bigger concern when implementing quicksort is the stack space usage. In order to have a limit on the stack space the compiler has to be able to do tail call optimization, having your sort crash depending on the ability of the compiler is probably worse than the quadratic performance of this case.